From: Manish Chopra <manish.chopra@qlogic.com>
Date: Thu, 2 Jan 2014 13:38:44 -0500
Subject: [PATCH 173/270] qlcnic: Fix resource allocation for TX queues
Patch-mainline: v3.13-rc8
Git-commit: f3e3ccf83bab261c5b55623bd3e9d1147b1c2e19
References: bsc#909350 FATE#317546

o TX queues allocation was getting distributed equally among all the
  functions of the port including VFs and PF. Which was leading to failure
  in PF's multiple TX queues creation.

o Instead of dividing queues equally allocate one TX queue for each VF as VF
  doesn't support multiple TX queues.

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
Acked-by: Gary Ching-Pang Lin <glin@suse.com>
---
 drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
@@ -76,7 +76,6 @@ static int qlcnic_sriov_pf_cal_res_limit
 	num_vfs = sriov->num_vfs;
 	max = num_vfs + 1;
 	info->bit_offsets = 0xffff;
-	info->max_tx_ques = res->num_tx_queues / max;
 
 	if (qlcnic_83xx_pf_check(adapter))
 		num_macs = 1;
@@ -91,6 +90,7 @@ static int qlcnic_sriov_pf_cal_res_limit
 		temp = num_macs * num_vfs * QLCNIC_SRIOV_VF_MAX_MAC;
 		temp = res->num_rx_mcast_mac_filters - temp;
 		info->max_rx_mcast_mac_filters = temp;
+		info->max_tx_ques = res->num_tx_queues - sriov->num_vfs;
 
 	} else {
 		id = qlcnic_sriov_func_to_index(adapter, func);
@@ -103,6 +103,7 @@ static int qlcnic_sriov_pf_cal_res_limit
 		info->max_tx_mac_filters = num_macs;
 		temp = num_macs * QLCNIC_SRIOV_VF_MAX_MAC;
 		info->max_rx_mcast_mac_filters = temp;
+		info->max_tx_ques = QLCNIC_SINGLE_RING;
 	}
 
 	info->max_rx_ip_addr = res->num_destip / max;
